index.js

```diff
@@ -0,0 +1,53 @@
+const through2Concurrent = require('through2-concurrent');
+
+class MongoToSqs {
+
+  constructor(options) {
+    this.cursor = options.cursor; // db.collection(name).find()
+    this.sqs = options.sqs; // new AWS.SQS()
+    this.queueUrl = options.queueUrl; // https://sqs.us-east-1.amazonaws.com/000000000000/my-queue
+    this.formatPayload = options.formatPayload || (doc => doc);
+    this.concurrency = options.concurrency || 2500;
+    // TODO: this.sqsBatchSize - instead of sending every sqs doc individually
+  }
+
+  sendToQueue(payload) {
+    return this.sqs.sendMessage({
+      MessageBody: JSON.stringify(payload),
+      DelaySeconds: 0,
+      QueueUrl: this.queueUrl
+    }).promise();
+  }
+
+  start() {
+    let count = 0;
+    return new Promise((resolve, reject) => {
+      this.cursor.pipe(through2Concurrent.obj({
+        maxConcurrency: this.concurrency
+      }, (doc, encoding, done) => {
+        count += 1;
+        const payload = this.formatPayload(doc);
+        this.sendToQueue(payload)
+          .then(() => {
+            count -= 1;
+            done();
+          })
+          .catch(err => {
+            count -= 1;
+            console.log('sendToQueue:', doc, err);
+            done();
+          });
+      }));
+      // done reading from mongo collection
+      this.cursor.on('end', () => {
+        // wait until all docs have been processed
+        setInterval(() => {
+          if (count === 0) resolve();
+        }, 1000);
```

Comment:
`setInterval` feels hacky and `count` could be in a race condition where it ends up being `0` but there was a lagged request that bumps it back to `1`. 
Could it be resolved by pushing promises to an array and wait for `Promise.all`?

Reply:
I agree setInterval is a bit hacky, but I think it's better than nextTick in this case. I can't use `Promise.all` because it runs out of memory after 1M docs or so.

You may be right about a race condition existing and that it works because it always takes less than 1000ms to iterate the last batch of mongo docs. Let me take a closer look at that.

Reply:
What about pushing to a pending-requests stack. Whenever, they complete, pop them from the stack so its not infinitely growing. Then on `cursor.end` check the stack

Reply:
I removed the race condition. Thanks @aoxrud-ww and @Jin-WW.